lib/util/github/graphql/cache-adapters/abstract-cache-adapter.ts

```diff
@@ -0,0 +1,160 @@
+import { DateTime } from 'luxon';
+import type {
+  GithubDatasourceItem,
+  GithubGraphqlCacheAdapter,
+  GithubGraphqlCacheRecord,
+} from '../types';
+import { isDateExpired } from '../util';
+
+/**
+ * Cache adapter handles the caching Github GraphQL items
+ * and reconciling them with newly obtained from paginated queries.
+ */
+export abstract class AbstractGithubGraphqlCacheAdapter<
+  GithubItem extends GithubDatasourceItem
+> implements GithubGraphqlCacheAdapter<GithubItem>
+{
+  /**
+   * Time period during which a package can be evicted from cache.
+   */
+  protected static readonly packageUnstableDays = 7;
+
+  /**
+   * Time period after which a cache record is considered expired.
+   */
+  protected static readonly cacheTTLDays = 30;
+
+  /**
+   * The time which is used during single cache access cycle.
+   */
+  protected readonly now = DateTime.now();
+
+  /**
+   * Set of all versions which were reconciled
+   * during the current cache access cycle.
+   */
+  private readonly reconciledVersions = new Set<string>();
+
+  /**
+   * These fields will be persisted.
+   */
+  private items: Record<string, GithubItem> | undefined;
+  protected createdAt = this.now;
+  protected updatedAt = this.now;
+
+  constructor(
+    protected readonly cacheNs: string,
+    protected readonly cacheKey: string
+  ) {}
+
+  /**
+   * Load data previously persisted by this adapter
+   * for given `cacheNs` and `cacheKey`.
+   */
+  private async getItems(): Promise<Record<string, GithubItem>> {
+    if (this.items) {
+      return this.items;
+    }
+
+    let result: GithubGraphqlCacheRecord<GithubItem> = {
+      items: {},
+      createdAt: this.createdAt.toISO(),
+      updatedAt: this.updatedAt.toISO(),
+    };
+
+    const storedData = await this.load();
+    if (storedData) {
+      const cacheTTLDuration = {
+        days: AbstractGithubGraphqlCacheAdapter.cacheTTLDays,
+      };
+      if (!isDateExpired(this.now, storedData.createdAt, cacheTTLDuration)) {
+        result = storedData;
+      }
+    }
+
+    this.createdAt = DateTime.fromISO(result.createdAt);
+    this.updatedAt = DateTime.fromISO(result.updatedAt);
+    this.items = result.items;
+    return this.items;
+  }
+
+  /**
+   * If a package version was published less than `packageUnstableDays` ago,
+   * then it is considered unstable and can be evicted from cache.
```

Comment:
Can you describe more about this? What's the importance of "unstable"?

BTW we may want to use a different term, seeing as stable/unstable have very specific meanings for releases

Reply:
Let's say you publish your package today. You have 7 days to remove or modify your package, so that cache will be updated. Now I started to think of removing this option and extend this period to `cacheTTLDays`. So fresh packages can be updated for 30 days, and after 30 we reset entire cache for this package anyway.